### ISSUE TITLE
Video Player

### ISSUE THREAD

**Issue by TheSuunny** [AvaloniaUI/Avalonia#2571]
Hi, is there any way to play the video through Avalonia?

I found a project, but it works only on Windows, i would like cross-platform.

**Comment by WamWooWam** [comment]
To do this in Avalonia itself seems not-super-feasible, because you'd have to either include some kind of cross platform media player (VLC?) with Avalonia, or hook the low level media APIs in every single supported operating system, which wouldn't be super easy as far as i can tell?

**Comment by jmacato** [comment]
**Disclaimer: The following commentary is my personal opinion regarding this matter. Any opinion I say in this space is not reflective of The Avalonia Project's current/future plans or endeavors and is also not representative of any consensus or opinion of any other core maintainer.**

I think this won't be in our focus for the time-being, as @WamWooWam said, it's a bit tricky to coordinate a sane approach towards media decoding. First off we have a crippling bug for this kind of applications (#2244) because we need to shuttle image data to-&-fro native codecs to managed .NET space. Wrappers like [FFMpeg.Autogen](https://github.com/Ruslan-B/FFmpeg.AutoGen) will definitely ease it up but for now it'd have to take a backseat.

**Comment by Sergey-Terekhin** [comment]
May be this project can be ported to Avalonia: https://github.com/unosquare/ffmediaelement - it already has wrappers around ffmpeg and some WPF controls which relatively easy can be ported

**Comment by jmacato** [comment]
@Sergey-Terekhin yes I've been aware of that project even before I got into Avalonia, but there are still some showstopper bugs (#2244) that we need to address first. Besides porting that is not for the faint-hearted, trust me coz i tried before and i judged that we're better off learning how it interacts with ffmpeg.autogen instead of porting it outright.

**Comment by shawnallen85** [comment]
https://github.com/videolan/libvlcsharp/pull/173

Started working on this and hopefully it gets merged in. I've tested it on Windows, MacOS, and Linux (only Ubuntu so far) and it works pretty well.

**Comment by simonmurrell** [comment]
@shawnallen85 Is there a way to plug your video extension into an Avalonia UI application?

**Comment by shawnallen85** [comment]
Sorry for the delay -- this is what I have in my code for using the video extension:

In the XAML:

        <Panel>
            <controls:Video MediaPlayer="{Binding MediaPlayer}" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" />
        </Panel>

In the view model:

I have a property for the MediaPlayer (referenced above in the XAML):

public MediaPlayer MediaPlayer { get; }

Then in my functions I control the media player -- that being I start/stop/change the media, etc.

I hope that helps -- let me know.

Happy Thursday!
Shawn

**Comment by simonmurrell** [comment]
Do you have a simple demo with this in?

> Sorry for the delay -- this is what I have in my code for using the video extension:
> 
> In the XAML:
> 
> ```
>     <Panel>
>         <controls:Video MediaPlayer="{Binding MediaPlayer}" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" />
>     </Panel>
> ```
> 
> In the view model:
> 
> I have a property for the MediaPlayer (referenced above in the XAML):
> 
> public MediaPlayer MediaPlayer { get; }
> 
> Then in my functions I control the media player -- that being I start/stop/change the media, etc.
> 
> I hope that helps -- let me know.
> 
> Happy Thursday!
> Shawn
> 
> Edit:
> 
> I just noticed this, but my code is referencing the control before it was integrated into the LibVLC project officially -- controls:Video might actually be controls:VideoView.
> 
> I think the rest is the same.

**Comment by shawnallen85** [comment]
I don't -- I'll see if I can modify the app I was using this in to use the official control and push it to github in the next day or so.

**Comment by shawnallen85** [comment]
I figured if I didn't knock it out now it wouldn't happen soon :P.

https://github.com/shawnallen85/Homer

**Comment by simonmurrell** [comment]
Thanks. I will give it a whirl!

**Comment by zls3201** [comment]
I sucess.
you can read

https://github.com/donandren/vlcsharpavalonia/blob/master/src/LibVLCSharp.Avalonia/NativeVideoPresenter.cs

https://github.com/hudec117/Mpv.NET-lib-/blob/master/src/Mpv.NET.WPFExample/MainWindow.xaml.cs


My Code:

*.csproj add line
```
<PackageReference Include="Mpv.NET" Version="1.2.0" />
```

```cs
/// <summary>
    /// Avalonia VideoView for Windows, Linux and Mac.
    /// </summary>
    public class VideoView : NativeControlHost
    {
        private MpvPlayer player;
        IPlatformHandle CreateWin32(IPlatformHandle parent)
        {
            //var control = base.CreateNativeControlCore(parent);
            player = new MpvPlayer(parent.Handle, "win-x64/mpv-1.dll")
            {
                Loop = true,
                Volume = 100
            };
            player.Load("http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4");
            player.Resume();

            return parent;// new PlatformHandle(control.Handle, "HWND");

        }

        void DestroyWin32(IPlatformHandle handle)
        {
            player.Dispose();
            base.DestroyNativeControlCore(handle);
        }

        protected override IPlatformHandle CreateNativeControlCore(IPlatformHandle parent)
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return CreateWin32(parent);
            return base.CreateNativeControlCore(parent);
        }

        protected override void DestroyNativeControlCore(IPlatformHandle control)
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                DestroyWin32(control);

            base.DestroyNativeControlCore(control);
        }
    }
```
xaml
```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"			 
        xmlns:vlc="clr-namespace:ZTool.Common;assembly=ZTool"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="ZTool.Controls.VideoPlayerView">
	<vlc:VideoView  HorizontalAlignment="Stretch" VerticalAlignment="Stretch" />
</UserControl>

```

**Comment by bbhxwl** [comment]
Do any of you have a simple VLC demo? I want to use it on raspberry pie, otherwise I can only use pyqt5 it, but I'm familiar with C#

**Comment by Symbai** [comment]
@shawnallen85 There is currently no way to draw something on top of the mediaplayer (because it has no content property) but for LibVLCSharp's WPF implementation it is. Any chance this can be done for Avalonia as well?

**Comment by Geektoolkit** [comment]
Seconding this....I'd really really like to be able to draw on top of the video. Right now event trying to put something on top of it in Z order isn't working (I'm not using full screen mode).  I'd really like to be able to have info about the video thats playing show up in an avalonia window I control while the videos playing

**Comment by radiolondra** [comment]
For anyone is interested:

I created an open source repo on Github which tries to overcome some Avalonia LibVlcSharp issues and missed things.
Give a look to **YAMP** (**Y**et **A**nother **M**edia **P**layer) [here](https://github.com/radiolondra/YAMP).
I created also the setups (x86 & x64) to install YAMP and test it directly (version 0.1.10.220) [here](https://github.com/radiolondra/YAMP/releases).

It is obviously just a starting point. A lot of work is still needed to have YAMP stable.
Any help, contribution, discussion, improvement ideas or just one coffee together is welcome.
Many thanks for your attention.

**Comment by Symbai** [comment]
> I created an open source repo on Github which tries to overcome some Avalonia LibVlcSharp issues and missed things.

As far as I've seen on your project description it does not overcome any of the mentioned issues. It still needs an own window and you still cannot draw elements on top of it. Besides, it is not a control but an application which isn't really what people here are looking for (although they could learn/read the source code). Still thanks for sharing.

**Comment by radiolondra** [comment]
@Symbai Yes it doesn't solve the issues, just tries to live with them and have a decent result.

**Comment by Geektoolkit** [comment]
I've spent quite a few hours on this working on a raspberry pi, and here's what I've found (This is all raspberry pi specific, but I find it interesting because it may help others learn what I learned..which is that video is insanely complex)
1) Even VLC, which is the preferred player for a rapsberry pi bullseye version of Debian...has problems with overlaying...simply turning on subtitles or placing a window on top of it when it's not in full screen mode causes issue
2) Creating a window in avalonia and then calling 'activate' will bring it over the top of vlc, but again due to point 1...it causes it to perform horribly
3) If VLC is in full screen mode, nothing can be placed on top of it (avalonia or otherwise), but it performs great 
4) I and my team tried working around this by using the logo feature of VLC, and that cause perf to drop also
5)  Launching VLC 'windowed' and then trying to resize it using width/height command line args doesn't work (7 year old bug they say is fixed in 4.0...but I don't see an arm build and it's in alpha)
6) I tried using MPV and FFPlay...by using process.start and the correct command lines you can full screen launch them, but they also either fall over when a window is placed over them or aren't GPU accelerated on a pi (I hear I can build a version of them, but then there are reports bullseye broke it, etc)
7) To make windows transparent on debian using compmgr, use -o.0   when running it and then avalonia transparency works, otherwise you get white backgrounds (this is how you can get a window on top of a video

Note: On windows...this all works flawlessly for me. No perf issues with avalonia or vlc, and you can easily put a window on top of the video using Activate, with it transparent and everything working.  

I've tried the VLCcontrol (which I'll continue to use...it's very nice) for Avalonia built upon libvlc, as well as tried shell executiong VLC.

Those are my findings thought they'd be useful for others.  Radiondra if your MPV solution works on arm/debian I could give it a shot, but unless you're using a hardware accleerated version of MPV it wont do me or others much good that are doing this on arm/debian platforms like the pi4

**Comment by radiolondra** [comment]
@Geektoolkit YAMP doesn't use MPV, it uses LibVlc mediaplayer.
You did really an interesting job with Raspberry. I'm working with a couple of pi4 for a TinyML project and peraphs I could use one of them to try Vlc&avalonia things there and test. I'lllet you know. The only variable against this is time, which is too short. Anyway, I'll try for sure.

**Comment by radiolondra** [comment]
I created 2 repos on Github to show how to use **VLC MediaPlayer embedded** into Avalonia [Window](https://github.com/radiolondra/AvaVLCWindow)s and [UserControl](https://github.com/radiolondra/AvaVLCControl)s.

Still trying to find the best way to add a UserControl (i.e. media player buttons) on top of MediaPlayer.

The simplest and faster way is to handle a new window (already used in [YAMP](https://github.com/radiolondra/YAMP)), but I desire to find a more flexible and integrated way. 

Help, suggestions and comments are welcome (obviously).
Thanks.

**Comment by radiolondra** [comment]
**How to add a UserControl on top of VLC MediaPlayer**

I modified the [Window ](https://github.com/radiolondra/AvaVLCWindow) repository to show how to put a UserControl on top of MediaPlayer (VideoView).

To do this I added a Content property to VideoView which will create its own new top window and add the Content to this window.

You can create the Content as a new UserControl (for example, a MediaPlayer set of controls) and add this UserControl to the VideoView content, like in:

```
<vlcsharp:VideoView
   HorizontalAlignment="Stretch"
   VerticalAlignment="Stretch"
   x:Name="VideoViewer">

   <!-- Content -->
   <Panel
	VerticalAlignment="Bottom"
	Opacity="0.8"
	Background="Gray">
	<controls:PlayerControls/>
   </Panel>
</vlcsharp:VideoView>
```
In this repo I created a user control with 2 buttons: Play and Stop. 
The control position is at bottom and the buttons are shown using PointerEnter/PointerLeave events, with a simple opacity set to 0.8 (Enter) and to 0 (Leave). Too lazy to add some kind of animation to the UserControl.

I'll modify the [UserControl](https://github.com/radiolondra/AvaVLCControl) repository too, adding the same feature, even though you could do it by yourselves.

Remember that, even though it works, the code is very basic and could need still a bit of work to be perfect.

**Comment by radiolondra** [comment]
**EDIT:** 
Successfully tested on **Kubuntu 18.04**

![AvaVLCWindow-Kubuntux64](https://user-images.githubusercontent.com/20070559/161376109-6138f911-d9ac-4f23-bef2-cbc04ddb266e.png)

Successfully tested on **MacOS 10.13 (High Sierra)**

![AvaVLCWindow-MacOS](https://user-images.githubusercontent.com/20070559/161376137-ade56fd3-c2ff-4199-aeba-a2bdc0fc4474.png)

**Comment by radiolondra** [comment]
**Test on Raspberry Pi (Raspbian Stretch)**
926MB RAM

As @Geektoolkit said, generated on-top window has no transparency and opacity is not respected. Like @Geektoolkit, I tried every possible combinations of Background and TransparencyLevelHint without any success. Maybe Avalonia has some issues for linux-arm or maybe there are some other settings I don't know.

![AvaVLCWindow-RaspberryPi](https://user-images.githubusercontent.com/20070559/161395884-05fcd867-5527-459e-8cd7-198ff141b17a.png)

Having 926MB of RAM only and just 1 CPU, the video is played in jerks. Anyway, it works BUT the transparency and opacity issues.

**Comment by Geektoolkit** [comment]
To get opacity try xcommgr -o.0 to see if that helps with opacity. I got the pi to work using that command/setting

**Comment by SilkyFowl** [comment]
@radiolondra 
I can confirm that this approach works [just fine with Avalonia.FuncUI(F#)!](https://github.com/SilkyFowl/FuncUI.VideoPlayer/blob/master/src/VideoPlayer/VideoPlayer.fs) Thanks for sharing the idea!

**Comment by radiolondra** [comment]
## New test on Raspberry Pi (DietPi with MATE or Xfce)

### It works fine! 
**Transparency and Opacity are fully honored.**

In my opinion, the problem is not Avalonia vs Raspberry, the problem is what is installed or not in Raspberry.

With **DietPi OS and MATE (or Xfce) as X**, the sample works perfectly. 

As root, I installed `libx11-dev, libvlc-dev and vlc`, copied the published sample (built in Windows 10 with `dotnet publish -c release -f net5.0 -r linux-arm`), `chmod 777 the whole folder content` and ran the sample. **It has to be ran as root or with sudo**.

![rpi](https://user-images.githubusercontent.com/20070559/161716907-ab810f65-e1d6-4199-b931-0656faf649c8.png)
![rpi_000](https://user-images.githubusercontent.com/20070559/161716966-e66fee8f-d82b-47fd-9366-4d7da1130265.png)

**Comment by radiolondra** [comment]
> OMG this would be a game changer for those of us doing video on raspberry pis! I'll give this a try asap...that looks amazing! And from what I'm gathering, the video is playing without stuttering?!

The video is playing per-fec-tly! 

DietPi is lighter than whatever else Raspberry OS, but not less complete.

Be sure to install MATE or Xfce, I didn't test yet with the other X options.

Be prepared to swallow a couple of shirts before to have it installed in a clean way. I had to flash it 3 times (on the SD). 

At start, it uses the old text installation (it remembers the old RedHat way to install Linux) and it seems not so obvious. After some attempts (you have to manually configure the network too) you will thanks this kind of approach, as it's very powerful.

Remember to use "root" (default password is "dietpi") to install what you need (libx11-dev, libvlc-dev and vlc). Ther's no X pre-installed, so you have to choose which X you want and install it (MATE).

Once X is installed, if you have the default (text) login start MATE with normal "startx" command.

Now copy your published (full) project somewhere (I put mine in  root/Desktop), then add needed libraries above. I don't know if they are the ONLY libs needed, they were for my test project.

You should set exec rights to your project files (I used chmod 777 for all) and start your app as root (or using sudo).

Everything should work. 
Avalonia is perfect for Raspberry too.

**Comment by radiolondra** [comment]
**For anyone is interested**:

A sample to apply in the real world the changes I did for LibVLCSharp.Avalonia (**Unofficial** release [here](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial)).

I created **YAMP2** repository (**Y**et-**A**nother-**M**edia-**P**layer v2) [here](https://github.com/radiolondra/YAMP2) which fully uses the new LibVLCSharp.Avalonia.Unofficial lib.

To make tests, build it by yourselves and run.

Enjoy.

**Comment by radiolondra** [comment]
[The new release](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial.UCanvas) of the _Unofficial Avalonia LibVLCSharp_, allows to **display multiple draggable controls over the scene of a video played with LibVLCSharp** in an Avalonia app.

**For anyone is interested.**
Enjoy.

**Comment by radiolondra** [comment]
## All the links updated

### External Links
**VLC video player inside an Avalonia Window/UserControl embedding single static customizable control (e.g.player buttons)**
[LibVLCSharp.Avalonia.Unofficial](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial)

**VLC video player inside an Avalonia Window/UserControl embedding multiple draggable customizable controls (e.g.player buttons, images,...)**
[LibVLCSharp.Avalonia.Unofficial.UCanvas](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial.UCanvas)

**LibVLCSharp.Avalonia.Unofficial Samples**
[LibVLCSharp.Avalonia.Unofficial.Samples](https://github.com/radiolondra/libvlcsharp/tree/3.x/samples/LibVLCSharp.Avalonia.Unofficial.Samples)

**LibVLCSharp.Avalonia.Unofficial.UCanvas sample**
[LibVLCSharp.Avalonia.Unofficial.UCanvas.Samples](https://github.com/radiolondra/libvlcsharp/tree/3.x/samples/LibVLCSharp.Avalonia.Unofficial.UCanvas.Samples)

**YAMP 2 - Open source video player sample using LibVLCSharp.Avalonia.Unofficial library**
[Github sources](https://github.com/radiolondra/YAMP2)
[Watch YAMP2 in action](https://www.screencast.com/t/uTGBnpE6v)

**Comment by xiaokun** [comment]
The overlapped window which hosts the content will be disappeared if it lost focus on macOS 12.6.3. 

I changed some code (add a TextBox) in MainWindow.axaml to produce this problem:
`
<Grid RowDefinitions="*, 40"  ColumnDefinitions="*">		
       <controls:PlayerControl  Name="Player"/>
        <Grid Grid.Row="1" >
            <TextBox  />
        </Grid>
</Grid>
`
the overlapped content will disappear if your click on the TextBox. 

Maybe someone has solution. Thanks.

**Comment by clawton** [comment]
Anyone tried this on a DRM  framebuffer based configuration?

**Comment by timunie** [comment]
I'm going to move this issue into discussions as there is nothing to do on Avalonia code side
